### DISCUSSION TITLE
Service worker as middleware for backend search-index

### DISCUSSION THREAD

**Discussion by eklem** [fergiemcdowall/search-index#594] [Ideas]
To make it easier to create good frontends on search-index, I think we need a search-index service worker. Then frontend will only look at user interactions and drawing HTML from what comes from search-index. The service worker will be the middle-ware that takes user input, send it to search-index and returns data from search-index back to the frontend. Then we'll be able to do some heavy lifting indexing too, without interfering with the frontend too much.

**Comment by fergiemcdowall** [comment]
A cross-platform Service Worker for indexing is a really interesting idea. The background sync API shows a lot of promise, but is [STILL unavailable on Safari](https://caniuse.com/background-sync) (probably due to the threat the PWAs pose to the App Store)

**Comment by eklem** [comment]
Yes, think communication back and forth between GUI and SW + firing off search-index requests for the full API should be available. 

Safari may pick up speed in developing a modern browser since they are under some legal threat in the US. Either that or other browser engines allowed, I'm guessing. But the service worker is at least available. I'm mostly just thinking of "the browser is the server" rather than "offline first", but for many, the last one will be the important one.
